# Benchmarking Scripts
Python Bioblend scripts for automating tasks in Galaxy

## Prerequisites

The only Python requirement (so far) is the Bioblend library.

Create a Python virtual environment and install the *bioblend* and *pyyaml* libraries:

```
python3 -m venv .venv
source .venv/bin/activate
pip install --upgrade pip
pip install bioblend
pip install pyyaml
```

**NOTE:** On Linux and MacOS system you can set the executable bit on the `workflow.py`
file so you do not need to explicitly run the script with Python.

```
chmod +x workflow.py
./workflow.py help
```

### OPTIONS
```
    -k|--key GALAXY_API_KEY
        Specify the Galaxy API for the remote server
    -s|--server
        The URL for the remote Galaxy server
```
Both the Galaxy API key and Galaxy server URL can be specified in environment variables.

```
export GALAXY_API_KEY=<your api key>
export GALAXY_SERVER=https://benchmarking.usegvl.org/initial/galaxy
```



### COMMANDS

```
    wf|workflows
        List all public workflows and their inputs
    hist|histories
        List all public histories and their datasets
    st|status <invocation_id>
        If the invocation_id is specified then the invocation report for that workflow
        invocation is returned.  Otherwise lists all the workflow invocations on
        the server
    run <configuration.yml>
        Run the workflow specified in the configuration.yml file.
    version
        Print the version number and exit
    help|-h|--help
        Prints this help screen
```

When a workflow is run with the `run` command the invocation details will be saved to a JSON file with the invocation ID as the file name with a *.json* extension.  Use the invocation ID with the `st` (`status`) command to get detailed information about that invocation.

### EXAMPLES

    ./workflow.py run configs/paired-dna.yml
    ./workflow.py st da4e6f496166d13f

## Runtime Configuration

The runtime parameters for benchmarking runs are specified in a YAML configuration file.  The configuration file can contain more than one runtime configuration specified as a YAML list. This file can be stored anywhere, but several examples are included in the `config` directory. 

The YAML configuration for a single run looks like:

```
- workflow_id: d6d3c2119c4849e4
  output_history_base_name: RNA-seq
  reference_data:
    - name: Reference Transcript (FASTA)
      dataset_id: 50a269b7a99356aa
  runs:
    - history_name: 1
      inputs:
      - name: FASTQ RNA Dataset
        dataset_id: 28fa757e56346a34
    - history_name: 2
      inputs:
      - name: FASTQ RNA Dataset
        dataset_id: 1faa2d3b2ed5c436
```

- **workflow_id** 
  The ID of the workflow to run.
  
- **output_history_ base_name**  (optional)
  Name to use as the basis for histories created.  If the *output_history_base_name* is not specified then the  *workflow_id* is used.
  
- **reference_data** (optional)
  Input data that is the same for all benchmarking runs and only needs to be set once.  See the section on *inputs* below for a description of the fields

- **runs**
  Input definitions for a benchmarking run.  Each run defintion shoud contain:

  - **history_name** (optional) 
    The name of the history created for the output.  The final output history name is generated by concatenating the *output_history_base_name* from above and the *history_name*.  If the *history_name* is not specified an incrementing integer counter is used.
  - **inputs**
    The one or more input datasets to the workflow.  Each input specification consists of:
    1. **name** the input name as specified in the workflow editor
    2. **dataset_id** the History API ID as displayed in the workflow editor or with the `./workflow.py histories` command.

  
### Contributing

Fork this repository and then create a working branch for yourself from the `dev` branch. All pull requests should target  `dev` and not the `master` branch.

```bash
git clone https://github.com/ksuderman/bioblend-scripts.git
cd bioblend-scripts
git checkout -b my-branch
```

If you decide to work on one of the [issues](bioblend-scripts/issues) be sure to assign yourself to that issue to let others know the issue is taken.

